cupy/sorting/search.py

```diff
@@ -181,7 +182,128 @@ def where(condition, x=None, y=None):
     return _where_ufunc(condition.astype('?'), x, y)
 
 
-# TODO(okuta): Implement searchsorted
+# This is to allow using the same kernels for all dtypes, ints & floats
+# as nan is a special case
+_preamble = '''
+template<typename T>
+__device__ bool _isnan(T val) {
+    return false;
+}
+template<>
+__device__ bool _isnan<float16>(float16 val) {
+    return isnan(val);
+}
+template<>
+__device__ bool _isnan<float>(float val) {
+    return isnan(val);
+}
+template<>
+__device__ bool _isnan<double>(double val) {
+    return isnan(val);
+}
+template<>
+__device__ bool _isnan<const complex<double>&>(const complex<double>& val) {
+    return isnan(val);
+}
+template<>
+__device__ bool _isnan<const complex<float>&>(const complex<float>& val) {
+    return isnan(val);
+}
+'''
+
+
+_searchsorted_kernel = core.ElementwiseKernel(
+    'S x, raw T bins, int64 n_bins, bool side_is_right',
+    'U y',
+    '''
+    bool greater = (side_is_right ? x >= bins[n_bins-1] : x > bins[n_bins-1]);
+    if (_isnan<S>(x) && !side_is_right) {
+        int pos = n_bins;
+        for( ; pos > 0 && _isnan<T>(bins[pos-1]); pos--) {
```

Comment:
Suggestion for readability
```suggestion
        while (pos > 0 && _isnan<T>(bins[pos-1])) {
            --pos;
        }
```